astroimtools/cutout_tools.py

```diff
@@ -16,11 +16,267 @@
+    else:
+        if isinstance(image, fits.hdu.hdulist.HDUList):
+            image_hdu = image[image_ext]
+        elif isinstance(image, (PrimaryHDU, ImageHDU, CompImageHDU)):
+            image_hdu = image
+        else:
+            raise TypeError("Expected ImageHDU, HDUList, or file name. Got {0} instead".format(type(image)))
+        data = image_hdu.data
+        wcs = WCS(image_hdu.header)
+
+    # Calculate the pixel scale of input image:
+    pixel_scale_width = proj_plane_pixel_scales(wcs)[0] * u.Unit(wcs.wcs.cunit[0]) / u.pix
+    pixel_scale_height = proj_plane_pixel_scales(wcs)[1] * u.Unit(wcs.wcs.cunit[1]) / u.pix
+
+    # Check if `SkyCoord`s are available:
+    if 'ra' in catalog.colnames and 'dec' in catalog.colnames:
+        if 'x' in catalog.colnames and 'y' in catalog.colnames:
+            raise Exception("Ambiguous catalog: Both (ra, dec) and pixel positions provided.")
+        if catalog['ra'].unit is None or catalog['dec'].unit is None:
+            raise u.UnitsError("Units not specified for ra and/or dec.")
+        coords = SkyCoord(catalog['ra'], catalog['dec'], unit=(catalog['ra'].unit,
+                                                               catalog['dec'].unit))
+    elif 'x' in catalog.colnames and 'y' in catalog.colnames:
+        coords = SkyCoord.from_pixel(catalog['x'].astype(float), catalog['y'].astype(float), wcs, origin=origin)
+    else:
+        try:
+            coords = SkyCoord.guess_from_table(catalog)
+        except Exception as e:
+            raise e
+
+    # Figure out cutout size:
+    if 'cutout_width' in catalog.colnames:
+        if catalog['cutout_width'].unit is None:
+            raise u.UnitsError("Units not specified for cutout_width.")
+        if catalog['cutout_width'].unit == u.pix:
+            width = catalog['cutout_width'].astype(float)  # pix
+        else:
+            width = (catalog['cutout_width'] / pixel_scale_width).decompose().value  # pix
+    else:
+        raise Exception("cutout_width column not found in catalog.")
+
+    if 'cutout_height' in catalog.colnames:
+        if catalog['cutout_height'].unit is None:
+            raise u.UnitsError("Units not specified for cutout_height.")
+        if catalog['cutout_height'].unit == u.pix:
+            height = catalog['cutout_height'].astype(float)  # pix
+        else:
+            height = (catalog['cutout_height'] / pixel_scale_height).decompose().value  # pix
+    else:
+        raise Exception("cutout_height column not found in catalog.")
+
+    # Do not rotate if column is missing.
+    if 'cutout_pa' in catalog.colnames and not suppress_rotation:
+        if catalog['cutout_pa'].unit is None:
+            raise u.UnitsError("Units not specified for cutout_pa.")
+        apply_rotation = True
+    else:
+        apply_rotation = False
+
+    # Sub-directory, relative to working directory.
+    if to_fits:
+        path = output_dir
+        if not os.path.exists(path):
+            os.mkdir(path)
```

Comment:
definitely should check the mkdir works.  Many times problems with upstream directory not existing or permissions. I would check for both of these types of exceptions.

Reply:
Ah, another question. If the path exists AND contains something then what happens?  What happens if I run this twice in a row with two different input but the same output directory?  If overwrite is True, and there are fewer files in the second run, will there be extraneous files in the output directory?  (make sense?)